### ISSUE TITLE
Zipf's law/ Zipf distribution

### ISSUE THREAD

**Issue by abhaydhiman** [sympy/sympy#20079]
Zipf's law is an empirical law formulated using mathematical statistics that refers to the fact that many types of data studied in the physical and social sciences can be approximated with a Zipfian distribution, one of a family of related discrete power-law probability distributions. **Zipf distribution is related to the zeta distribution but is not identical.**

read more about this over here:-
https://en.wikipedia.org/wiki/Zipf%27s_law

Now the Zipf law also uses harmonic number to compute its `PDF, MGF, and CF` so it's really necessary to add functionality for harmonic number if it is not already available in sympy something like this can be used:-
```
# harmonic number required to compute pdf for zipf's distribution/ zip law
def harmonic_number(n, m):
    number = S.Zero
    if n >= S.One and m >= S.Zero:
        number = harmonic_number(n-1, m) + (1/ n**m)
    return number
```

moving on further its pdf can also be computed like:-
```
def pdf(k, s, N):
    if s >= S.Zero and N >= S.One:
        return (1/ k**s) / (harmonic_number(N, s))
```

all of this then further be compiled in a separate class for `zipf distribution` along with its related functionalities.
and also I guess its good to add under sympy specifically in `drv_types` as it is identical to zeta distribution.

**Comment by Maelstrom6** [comment]
Almost about done with the PR. Could you please provide some sources and/or examples for the values of the pdf, expectation and variance of this distribution with sample parameters? It will be used for the examples and testing.

One last thing, could you explain why it is different from the zeta distribution? 
Is it something like `Zeta('x', s) == Zipf('x', s, oo)`?

**Comment by abhaydhiman** [comment]
> Is it something like `Zeta('x', s) == Zipf('x', s, oo)`?

well, I don't know that either this is valid or not but if it is then it might be an approximation of `zeta` distribution.

**Comment by naveensaigit** [comment]
I went through the [Zipf-Mandelbrot Distribution](https://en.wikipedia.org/wiki/Zipf%E2%80%93Mandelbrot_law) and it looks like the Zipf distribution is a special case of Zipf-Mandelbrot (q=0). Would it be better to implement this too?

**Comment by Maelstrom6** [comment]
The 3-parameter harmonic function I don't think is implemented in SymPy yet. It doesn't look too difficult to add but I have no knowledge of these harmonic functions. If the 3-parameter version is implemented, I can add the Zipf-Mandelbrot distribution no problem.

**Comment by oscarbenjamin** [comment]
Given this and #20100 I wonder if it makes more sense for the stats module to make it easier for users to create arbitrary probability distributions from state space and PMF/pdf rather than adding a new class for every possible named distribution.

**Comment by abhaydhiman** [comment]
> The 3-parameter harmonic function I don't think is implemented in SymPy yet. It doesn't look too difficult to add but I have no knowledge of these harmonic functions.

It can be done by a recursive function easily and I'll provide the code for harmonic functions but finding a good algorithm might need some time. 
But now the point is that do we have to add both `Zipf law` as well as `Zipf-Mandelbrot Distribution`?

What did you say @sylee957?

**Comment by Smit-create** [comment]
> Given this and #20100 I wonder if it makes more sense for the stats module to make it easier for users to create arbitrary probability distributions

This is possible presently.
https://github.com/sympy/sympy/blob/7437631ba09b240c0647129bde8896579c17ad58/sympy/stats/__init__.py#L63-L82

**Comment by Maelstrom6** [comment]
> This is possible presently.

Maybe we should provide an easy interface for the user to also input precomputed characteristic functions, cdfs and such. The pdf, cdf, mgf and cf all uniquely characterize a distribution if the function exists.

**Comment by oscarbenjamin** [comment]
Perhaps there could be separate Basic classes like PDFDistribution, MGFDistribution, etc representing a distribution defined from whichever feature.

Alternatively are there user-level functions for converting between those representations?

**Comment by Maelstrom6** [comment]
>Perhaps there could be separate Basic classes like PDFDistribution, MGFDistribution, etc representing a distribution defined from whichever feature.

I'm thinking something more along the lines that the user can input whatever they want and if they input more than one function, the extra functions are all cached for later use.

> Alternatively are there user-level functions for converting between those representations?

As far as I'm aware, users will either have to manually differentiate, perform a Laplace or Fourier transform to change between representations.

**Comment by sylee957** [comment]
In my opinion, although adding a nontrivial distribution function that can extend `ContinuousDistributionHandmade` can be useful,
But I'd also want to see more algebras over the objects such that it can make this way nontrivial.

For example, Wolfram may still have more trivial distribution functions than sympy, but I also see that they have implemented a lot of algebra between distributions such that it gives another predefined distribution objects than stuff like `ProbabilityDistribution` computed with scratch, which is one of the cool features I expect to see.

However, although a lot of algebraic relations are known for guassian, bernoulli, ... (So it gives a clear advantage not to use custom distribution for them), some functions may not have such rich features, so maybe it can be where a custom distribution function can be used.

**Comment by Smit-create** [comment]
> The pdf, cdf, mgf and cf all uniquely characterize a distribution if the function exists.

Currently, `sympy` computes all other functions using `pdf` by itself. If we add CDF, mgf, cf, etc. it will decrease the computation cost for sure while calculating `probability`, `expectation`, or the function itself.  If the user wants to add CDF, mgf, etc. then it can be just created following the way the current distributions are defined. For example,
https://github.com/sympy/sympy/blob/7437631ba09b240c0647129bde8896579c17ad58/sympy/stats/crv_types.py#L2847-L2871

**Comment by oscarbenjamin** [comment]
Yes, exactly. There is no reason users can't define their own probability distribution classes. Adding distributions seems to be a popular way to contribute to the stats library but I think that means it ends up being lots of code of and docs for lots of obscure distributions. I expect that it would be more useful for users to be able to:

1. Do more interesting things with more common distributions.
2. Solve problems involving arbitrary user-defined distributions

**Comment by Maelstrom6** [comment]
> lot of algebraic relations are known for guassian, bernoulli, ...

> Do more interesting things with more common distributions.

How would we go about this? 
For example, any linear combination of normals is normal. Would we make a function like `simplify_pspace` and call it at the start of every `E` or `P` call? I couldn't find out how to reliably override `__add__`, `__mul__`, etc. so I think we would need to do it outside the `NormalDistribution` class for this example.

**Comment by sylee957** [comment]
One idea would be introducing objects like `TransformedDistribution` in mathematica
https://reference.wolfram.com/language/guide/DerivedDistributions.html

**Comment by oscarbenjamin** [comment]
I think that the most important thing would be to start by thinking about what useful things users might want do with a symbolic stats module and then think about how to facilitate that. I'm not particularly involved in the stats module but I watch the discussion on issues and PRs and I don't have a sense of what the module is really used for.

**Comment by asmeurer** [comment]
Harmonic numbers are implemented in `harmonic`. Is the 3-argument form used generally? If so, it can be added to harmonic, otherwise, we should just implement it using the summation. The Wikipedia article for harmonic numbers doesn't seem to mention the 3-argument form. 

I agree that it should be easy for users to define their own distributions, but I also think that notable distributions should be put in SymPy. Zipf is notable in my opinion (at least I've heard of Zipf before, and I haven't heard of most of the ones that are in sympy.stats). 

The original design of the stats module is that distributions fallback to other things like integrate() to compute things. If we can define some distributions as transformations of other distributions and still have all the functions like expectation, etc. be computable, that would definitely simplify the code. For example, lognormal is currently defined independently, but is it possible to define it as literally exp(Normal), and have the pdf, etc. just be computed from that? Quite a few distributions are defined in this way, so if something like that can work, it would simplify a lot of code.

**Comment by naveensaigit** [comment]
> One idea would be introducing objects like `TransformedDistribution` in mathematica

How do you think it can be implemented? Using characteristic functions?

**Comment by Smit-create** [comment]
> I think that the most important thing would be to start by thinking about what useful things users might want do with a symbolic stats module and then think about how to facilitate that

Yes, I'd try to compile a list of issues and basic necessities that are required to be present in stats. In my opinion, we should avoid adding more distributions and focus on improving present features available in stats.

**Comment by thambi999** [comment]
U all not understand my mind dustb, can't to be thinking much, please settle fr meOn 18 Sep 2020 22:10, "S.Y. Lee" <notifications@github.com> wrote:
Actually, I don't think that sympy needs TransformedDistribution objects if we can already build arbitrary expressions with RandomSymbol
However, there should be a way to rewrite exp(Normal('X', 0, 1)) to LogNormal('Y', 0, 1)

**Comment by oscarbenjamin** [comment]
> Actually, I don't think that sympy needs `TransformedDistribution` objects if we can already build arbitrary expressions with RandomSymbol

I wonder if we should make distribution objects more of a central part of the usage of the stats module. It doesn't make much sense to simplify `exp(X)` to `Y` where `Y` has a different distribution. How would we decide what name to give `Y` and how would users query what distribution it had? Instead we could have a function for querying the distribution of an expression e.g.:
```python
>>> X = Normal('X', 0, 1)
>>> expr = 2*X + 1
>>> expr
2*X + 1
>>> distribution(expr)
TransformedDistribution(2*X + 1, (X, NormalDistribution(0, 1)))
>>> simplify(_)
NormalDistribution(1, 2)
```
Something analogous to `TransformedDistribution` is needed if working explicitly in distributions rather than expressions.

**Comment by sylee957** [comment]
It could be something like `TransformedRV('Y', 2*X + 1, (X, NormalDistribution(0, 1)))`

**Comment by sylee957** [comment]
However, I'm afraid that any approach like that will make the Random expression an obsolete idea, because you would be able to do do anything with TransformedDistribution that you can do with random expressions.

**Comment by naveensaigit** [comment]
> However, I'm afraid that any approach like that will make the Random expression an obsolete idea, because you would be able to do do anything with TransformedDistribution that you can do with random expressions.

If that's the case, we would have to override the +,-,/,* operators for the distribution classes and then, add a `is_random` check in every function sympy has to offer and simplify the expression of the distribution using its inverse . It is a lot of work!

**Comment by oscarbenjamin** [comment]
> > However, I'm afraid that any approach like that will make the Random expression an obsolete idea, because you would be able to do do anything with TransformedDistribution that you can do with random expressions.
> 
> If that's the case, we would have to override the +,-,/,* operators for the distribution classes and then, add a `is_random` check in every function sympy has to offer and simplify the expression of the distribution using its inverse . It is a lot of work!

No we wouldn't use operators like `+` with distributions. The point is just that you could query the distribution of a n expression involving a random variable which would generically be a TransformedDistribution:
```
>>> probability_distribution(exp(X))
TransformedDistribution(exp(x), (X, NormalDistribution(0, 1)))
```
You could then simplify that:
```
>>> simplify(_)
LogNormalDistribution(0, 1)
```
Or you could have rewrite rules and other things.

It doesn't make sense to simplify the random expression like:
```
>>> simplify(exp(X))
Y
```
because the user doesn't have any way of understanding what that means.

**Comment by naveensaigit** [comment]
> `>>> simplify(_)`
`LogNormalDistribution(0, 1)`

Do you have any idea how this can work?

**Comment by oscarbenjamin** [comment]
First there would need to be distribution objects for these things and a way to obtain the distribution from an expression. Then we can add `_eval_simplify` to `TransformedDistribution`.

**Comment by sylee957** [comment]
First, there should be a good survey of how the distributions we already have can relate each other with transformations.

**Comment by oscarbenjamin** [comment]
There should also be rewrite rules e.g.:
```python
>>> LogNormalDistribution(0, 1).rewrite(NormalDistribution)
TransformedDistribution(exp(x), (x, NormalDistribution(0, 1)))
```

**Comment by Maelstrom6** [comment]
> First, there should be a good survey of how the distributions we already have can relate each other with transformations.

The following pdf on page 26 (labeled page 20) has a graph of a few common transformations.
[2002-formulae-and-tables.pdf](https://github.com/sympy/sympy/files/5269238/pdfslide.net_2002-formulae-and-tables.pdf)

Some links if the attachment doesn't work:
[1](https://pdfslide.net/download/link/2002-formulae-and-tables)
[2](https://www.actuaries.org.uk/studying/prepare-your-exams/formulae-and-tables)
[3](https://www.scribd.com/doc/96504853/2002-Formulae-and-Tables)

**Comment by czgdp1807** [comment]
> There should also be rewrite rules e.g.:
> 
> ```python
> >>> LogNormalDistribution(0, 1).rewrite(NormalDistribution)
> TransformedDistribution(exp(x), (x, NormalDistribution(0, 1)))
> ```

This would be an awesome feature. Opening a new issue for it.
